### ISSUE TITLE
services definitions on OS X

### ISSUE THREAD

**Issue by nhooyr** [NixOS/nixpkgs#21252]
There are no launchd services for servers such as postgres on macOS, you have to manually start/stop postgres yourself. While it is trivial to create a script/alias to do this or even write the service definition yourself, it would be nice if nix could do it for you.

#5246 was opened to address this but it was abandoned.

**Comment by LnL7** [comment]
I've been working working on a project that tries some of the parts of nixos to darwin. At the moment it's a separate project because I'm not sure yet how to integrate it with the existing nixos modules, but there should be very little difference for the parts that overlap.

Feedback and help are appreciated to make it easier to get started with it.
https://github.com/LnL7/nix-darwin

**Comment by LnL7** [comment]
@copumpkin do you have any thoughts on this?

**Comment by globin** [comment]
ping @periklis, he had some thoughts on this, too.

**Comment by periklis** [comment]
Way months back, when i was doing my first baby steps with nixpkgs, i tried to create a PR for a new output type called "plist" to adress this issue. My assumptions were:
- A lot of projects deliver in the first place sane default plists, so use them.
- Make the usage of a plist an optional part of the derivation outputs,
   e.g. `outputs = ["lib", "dev"] ++ stdenv.lib.optional (stdenv.isDarwin) ["plist"];`
- If the default plist needed modifications, we could use an extra attribute `plist =''...'';`
- Make ouput type and attr plist only available in the stenv.darwin.
- Print out instructions for `launchctl load /nix/store/long-hash-name-version-plist/org.company.name.plist` for the user after installPhase.

However, i failed to produce the PR at that time, because of lack of knowledge of the internals of stdenv.darwin. I got lost in the recursion of `stageFun` calls at that time, when i tried to inject just a custom `build-plist.sh`  to extend the `defaultNativeBuildInputs` of the initial `stdenv/generic/default.nix`.

In summary, this just an abstraction of the approach currently found in packages like xquatz, that produce plist files in the derivation's attrset and reference to the contents during the phases.

**Comment by LnL7** [comment]
I might be bias, but I'm not sure how useful a plist output is in practice since there's no way to customise it. Using the module system makes more sense to me and has some nice advantages like type checking and nice descriptions for all of the [launchd options](https://github.com/LnL7/nix-darwin/blob/master/modules/launchd/launchd.nix).

**Comment by periklis** [comment]
@LnL7 Sure you are right, but modules are the heavy-lifting way. I could not foresee that nix-darwin would be possible at that moment. However, i having a plist attribute which anybody can override would do the same all the launchd options, but this is the dirty way.

**Comment by mmahut** [comment]
Are there any updates to this issue, please?

**Comment by nhooyr** [comment]
Going to close as it has been a long time and I'm not sure if this is relevant anymore.
